src/AlignLoads.cpp

```diff
@@ -0,0 +1,196 @@
+#include "AlignLoads.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Scope.h"
+#include "Bounds.h"
+#include "ModulusRemainder.h"
+#include "Simplify.h"
+namespace Halide {
+namespace Internal {
+using std::vector;
+
+class AlignLoads : public IRMutator {
+public:
+    AlignLoads(Target t, int vs) : target(t), required_alignment(vs) {}
+private:
+    Target target;
+    // The desired alignment of a vector load.
+    int required_alignment;
+    // Alignment info for variables in scope.
+    Scope<ModulusRemainder> alignment_info;
+    using IRMutator::visit;
+    ModulusRemainder get_alignment_info(Expr e) {
+        return modulus_remainder(e, alignment_info);
+    }
+    int natural_vector_lanes(Type t) {
+        return required_alignment / t.bytes();
+    }
+
+    Expr concat_and_shuffle(Expr vec_a, Expr vec_b, std::vector<Expr> &indices) {
+        Type dbl_t = vec_a.type().with_lanes(vec_a.type().lanes() * 2);
+        Expr dbl_vec = Call::make(dbl_t, Call::concat_vectors, { vec_a, vec_b }, Call::PureIntrinsic);
+        std::vector<Expr> args;
+        args.push_back(dbl_vec);
+        for (int i = 0; i < (int) indices.size(); i++) {
+            args.push_back(indices[i]);
+        }
+        return Call::make(vec_a.type(), Call::shuffle_vector, args, Call::PureIntrinsic);
+    }
+    Expr concat_and_shuffle(Expr vec_a, Expr vec_b, int start, int size) {
+        std::vector<Expr> args;
+        for (int i = start; i < start+size; i++) {
+            args.push_back(i);
+        }
+        return concat_and_shuffle(vec_a, vec_b, args);
+    }
+
+    bool get_alignment_check(const Ramp *ramp, int host_alignment, Type t, int *lanes_off) {
+        // If the start of the buffer is not aligned, we conservatively assume unaligned.
+        // This way, we can restrict alignment detection to the index, i.e the ramp itself.
+        if (host_alignment % required_alignment) return false;
+        int reqd_alignment_lanes = required_alignment / t.bytes();
+        return reduce_expr_modulo(ramp->base, reqd_alignment_lanes, lanes_off, alignment_info);
+    }
+    void visit(const Load *op) {
+        debug(4) << "AlignLoads: Working on " << (Expr) op << "..\n";
+        Expr index = mutate(op->index);
+        if (op->type.is_vector()) {
+            bool external = op->image.defined();
+            if (external) {
+                debug(4) << "AlignLoads: Not dealing with an external image\n";
+                debug(4) << (Expr) op << "\n";
+                expr = op;
+                return;
+            } else {
+                const Ramp *ramp = index.as<Ramp>();
+                const IntImm *stride = ramp ? ramp->stride.as<IntImm>() : NULL;
+                // We will work only on natural vectors supported by the target.
+                if (ramp->lanes != natural_vector_lanes(op->type)) {
```

Comment:
@dsharletg : To comment on 1 above, I think breaking up non-native loads, regardless of stride, into native loads and mutating them should be enough. The rest of the visit function takes care of native loads of strides 1 and 2. I think we should just break them down into native loads and concat them. Am I missing something here?